tensorflow/core/common_runtime/gpu/gpu_device.cc

```diff
@@ -170,15 +170,36 @@ class EigenGpuStreamDevice : public ::Eigen::StreamInterface {
       LogMemory::RecordRawDeallocation(operation_, step_id_, buffer, allocator_,
                                        true);
     }
+    // Allocators that use CUDA API calls (such as cuMemFree or cuMemFreeAsync)
+    // cannot be used inside stream callbacks. For those allocators, we use
+    // the callback only for logging purposes, and call allocator->DeallocateRaw
+    // immediately (synchronizing first if necessary).
+    const bool allocator_is_callback_safe = allocator_->IsSafeInGpuCallback();
     AsyncFreeData* afData =
-        new AsyncFreeData(allocator_, buffer, operation_, step_id_);
+        new AsyncFreeData(allocator_, buffer, operation_, step_id_,
+                          /*do_logging_only=*/!allocator_is_callback_safe);
 #if GOOGLE_CUDA
     cudaError_t err = cudaStreamAddCallback(*stream_, asyncFree, afData, 0);
```

Comment:
I don't understand why a callback was ever needed, even before this PR. Normal TF kernels call AllocateRaw() to get temp memory, enqueue a CUDA kernel, then call DeallocateRaw. This means sense since the order of CUDA kernels run on the compute stream is the same as the order the TF OpKernels run on the host.

 Do you know why is it different for Eigen when Eigen internally allocates temp memory?

Reply:
I couldn't work that out either. Unless I'm missing something, calling AllocateRaw(), launching a kernel, and then immediately calling DeallocateRaw() is not safe if the allocator is not stream-ordered, and afaict BFCAllocator is not stream-ordered(?).

Reply:
The (AllocateRaw(), kernel_launch, DeallocateRaw()) pattern is safe for OpKernels for all allocators (including BFCAllocator, GPUcudaMallocAllocator, and GpuCudaMallocAsyncAllocator).

I think BFCAllocator is "stream-ordered" in the sense that the caller must ensure all kernel launches that use memory from the BFCAllocator must be on the compute stream (unless synchronization is used). When a kernel touches memory from AllocateRaw(), the host may have already called DeallocateRaw(). This is OK, because by time another kernel uses the same memory from a subsequence AllocateRaw(), the old kernel must have already finished since both kernels are on the compute stream.

GPUcudaMallocAllocator is not "stream-ordered" but cudaFree I think does a synchronization so it should also be safe.

Reply:
OK that makes sense (though I can't say I like it).
I don't know the answer to your original question though (why it's different for Eigen). Is there anyone who might know?

Reply:
I asked @ezhulenev, but he also doesn't know. I don't think anyone knows.

I'm guessing a callback is not needed, and I'm guessing the code comment above starting with "Eigen Ops directly allocate memory" is incorrect. Can you try never having a callback, and simply calling `DeallocateRaw` in this function? This would simplify the PR since you don't need to add new Allocator methods.

Free feel to modify this PR or create a new one. Also don't worry about testing this. I will run extensive internal tests before merging.

Also, do you know what ops were hanging? I don't know of any ops that use Eigen and have Eigen allocate temporary memory. Perhaps they were custom ops?

Reply:
I traced one example of this deallocation path to the Eigen sum reductions here:
https://github.com/tensorflow/tensorflow/blob/b89a2f2/tensorflow/core/kernels/xent_op.h#L91-L103
which makes sense to me as it is likely that the reduction implementation requires temporary memory.

I've created a new PR that calls DeallocateRaw immediately instead of from the callback here: https://github.com/tensorflow/tensorflow/pull/57841